File: sentinel/runtime/diagnostics.lua
-- sentinel/runtime/diagnostics.lua
-- SENT-6.9: Diagnostics System
-- ADR 002 §19, ADR 008 §12
-- Unified diagnostic format with severity levels and stage attribution

local Diagnostics = {}
Diagnostics.__index = Diagnostics

-- Diagnostic severity levels
Diagnostics.Severity = {
    ERROR = "ERROR",
    WARNING = "WARNING",
    INFO = "INFO",
}

-- Compilation stages (matches Rust Stage enum)
Diagnostics.Stage = {
    StructuralValidation = "Structural Validation",
    ReferenceResolution = "Reference Resolution",
    BlueprintExpansion = "Blueprint Expansion",
    DependencyResolution = "Operation Dependency Resolution",
    GoalCoverage = "Goal Coverage Validation",
    Optimization = "Cross-Operation Optimization",
    Lowering = "Lowering",
}

-- Create an error diagnostic
-- @param code string Error code (e.g., "C-7001")
-- @param stage string Stage that produced this diagnostic
-- @param message string Human-readable message
-- @param entity string|nil Optional entity reference (e.g., operation name)
-- @return table Diagnostic
function Diagnostics.error(code, stage, message, entity)
    return {
        severity = Diagnostics.Severity.ERROR,
        code = code,
        stage = stage,
        message = message,
        entity = entity,
    }
end

-- Create a warning diagnostic
-- @param code string Warning code (e.g., "C-7002")
-- @param stage string Stage that produced this diagnostic
-- @param message string Human-readable message
-- @param entity string|nil Optional entity reference
-- @return table Diagnostic
function Diagnostics.warning(code, stage, message, entity)
    return {
        severity = Diagnostics.Severity.WARNING,
        code = code,
        stage = stage,
        message = message,
        entity = entity,
    }
end

-- Create an info diagnostic
-- @param code string Info code (e.g., "C-7003")
-- @param stage string Stage that produced this diagnostic
-- @param message string Human-readable message
-- @param entity string|nil Optional entity reference
-- @return table Diagnostic
function Diagnostics.info(code, stage, message, entity)
    return {
        severity = Diagnostics.Severity.INFO,
        code = code,
        stage = stage,
        message = message,
        entity = entity,
    }
end

-- Create diagnostic with suggested fix
-- @param diagnostic table Base diagnostic
-- @param suggested_fix string Suggested fix text
-- @return table Diagnostic with suggested_fix
function Diagnostics.with_fix(diagnostic, suggested_fix)
    diagnostic.suggested_fix = suggested_fix
    return diagnostic
end

return Diagnostics

File: sentinel-compiler/crates/sentinel-compiler/src/diagnostics.rs
use serde::{Deserialize, Serialize};

/// Diagnostic severity level.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum Severity {
    Error,
    Warning,
    Info,
}

/// Compilation stage that produced the diagnostic.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum Stage {
    StructuralValidation,
    ReferenceResolution,
    BlueprintExpansion,
    DependencyResolution,
    GoalCoverage,
    Optimization,
    Lowering,
}

/// A single diagnostic message emitted by the compiler.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Diagnostic {
    pub severity: Severity,
    pub code: String,
    pub stage: Stage,
    pub message: String,
    pub entity: Option<String>,
    pub suggested_fix: Option<String>,
}

impl Diagnostic {
    pub fn error(code: impl Into<String>, stage: Stage, message: impl Into<String>) -> Self {
        Self {
            severity: Severity::Error,
            code: code.into(),
            stage,
            message: message.into(),
            entity: None,
            suggested_fix: None,
        }
    }

    pub fn warning(code: impl Into<String>, stage: Stage, message: impl Into<String>) -> Self {
        Self {
            severity: Severity::Warning,
            code: code.into(),
            stage,
            message: message.into(),
            entity: None,
            suggested_fix: None,
        }
    }

    pub fn info(code: impl Into<String>, stage: Stage, message: impl Into<String>) -> Self {
        Self {
            severity: Severity::Info,
            code: code.into(),
            stage,
            message: message.into(),
            entity: None,
            suggested_fix: None,
        }
    }

    pub fn with_entity(mut self, entity: impl Into<String>) -> Self {
        self.entity = Some(entity.into());
        self
    }

    pub fn with_fix(mut self, fix: impl Into<String>) -> Self {
        self.suggested_fix = Some(fix.into());
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_diagnostic_creation() {
        let d = Diagnostic::error("C-1001", Stage::StructuralValidation, "duplicate id");
        assert_eq!(d.severity, Severity::Error);
        assert_eq!(d.code, "C-1001");
        assert!(d.entity.is_none());
        assert!(d.suggested_fix.is_none());
    }

    #[test]
    fn test_warning_with_entity_and_fix() {
        let d = Diagnostic::warning("C-1008", Stage::StructuralValidation, "version mismatch")
            .with_entity("Profile 'Test'")
            .with_fix("Update schema_version");
        assert_eq!(d.severity, Severity::Warning);
        assert_eq!(d.entity.as_deref(), Some("Profile 'Test'"));
        assert_eq!(d.suggested_fix.as_deref(), Some("Update schema_version"));
    }

    #[test]
    fn test_info_diagnostic_creation() {
        let d = Diagnostic::info("C-7001", Stage::Lowering, "Profile lowered successfully");
        assert_eq!(d.severity, Severity::Info);
        assert_eq!(d.code, "C-7001");
    }

    #[test]
    fn test_all_stage_variants() {
        assert!(matches!(Stage::StructuralValidation, Stage::StructuralValidation));
        assert!(matches!(Stage::ReferenceResolution, Stage::ReferenceResolution));
        assert!(matches!(Stage::BlueprintExpansion, Stage::BlueprintExpansion));
        assert!(matches!(Stage::DependencyResolution, Stage::DependencyResolution));
        assert!(matches!(Stage::GoalCoverage, Stage::GoalCoverage));
        assert!(matches!(Stage::Optimization, Stage::Optimization));
        assert!(matches!(Stage::Lowering, Stage::Lowering));
    }
}